Docs troubleshooting for Starboard Operator (#1082)
aquasecurity/starboard: docs/operator/troubleshooting.md

# Troubleshooting the Starboard Operator

The Starboard Operator installs several Kubernetes resources into your Kubernetes cluster.

Here are the common steps to check whether the operator is running correctly and to troubleshoot common issue.

In addition to having a look at this section, you want to check [previous issues](https://github.com/aquasecurity/starboard/issues) to see if someone from the community had similar problems before.
Feel free to either [open an issue](https://github.com/aquasecurity/starboard/issues), reach out on [Slack](https://slack.aquasec.com), or post your questions in the [discussion forum.](https://github.com/aquasecurity/starboard/discussions)

## Installation

Make sure that the latest version of the Starboard Operator is installed inside of your Kubernetes cluster.

> For this, have a look at the installation [options.](./installation)

Reviewer: Actually we don't have docs/installation/index.md that describes different installation options. You'll get 404 - Not found error.

We can either link to particular installation option, e.g. ./installation/kubectl.md or create such overview page. Or remove this link altogether.

Author: but it links to the index.md within the operator, when I run the docs locally, it works? 
<img width="252" alt="Screenshot 2022-03-31 at 10 35 29" src="https://user-images.githubusercontent.com/33576047/161025216-af49a785-a40a-4a4a-9321-ad83a44fb365.png">

Reviewer: Don't you see this warning in the logs:

```
INFO     -  Documentation built in 2.13 seconds
INFO     -  [09:37:54] Serving on http://0.0.0.0:8000/starboard/
WARNING  -  [09:37:59] "GET / HTTP/1.1" code 302
WARNING  -  [09:37:59] "GET /versions.json HTTP/1.1" code 404
INFO     -  [09:37:59] Browser connected: http://localhost:8000/starboard/
WARNING  -  [09:38:03] "GET /versions.json HTTP/1.1" code 404
INFO     -  [09:38:03] Browser connected: http://localhost:8000/starboard/operator/troubleshooting/
WARNING  -  [09:38:13] "GET /starboard/operator/troubleshooting/installation HTTP/1.1" code 404
WARNING  -  [09:38:13] "GET /versions.json HTTP/1.1" code 404
```

Anyway, if you intend to link to index.md use the .md suffix explicitly in MkDocs markdown.

Author: oh

Reviewer: See https://www.mkdocs.org/user-guide/writing-your-docs/#linking-to-pages for more details:

![mkdocs_links](https://user-images.githubusercontent.com/1322923/161027519-68585559-dc58-4a9d-ae32-599323463fa3.png)

```
kubectl describe pod <POD-NAME> -n starboard-system
```

To check the logs, use the following command:
```
kubectl logs deployment/starboard-operator -n starboard-system
```

If your pod is not running, try to look for errors as they can give an indication on the problem.

If there are too many logs messages, try deleting the Starboard pod and observe its behaviour upon restarting. A new pod should spin up automatically after deleting the failed pod.

## ImagePullBackOff or ErrImagePull

Check the status of the Starboard Operator pod running inside of your Kubernetes cluster. If the Status is ImagePullBackOff or ErrImagePull, it means that the Operator either

* tries to access the wrong image
* cannot pull the image from the registry

Make sure that you are providing the right resources upon installing the Starboard Operator.

## CrashLoopBackOff

If your pod is in `CrashLoopBackOff`, it is likely the case that the pod cannot be scheduled on the Kubernetes node that it is trying to schedule on.
In this case, you want to investigate further whether there is an issue with the node. It could for instance be the case that the node does not have sufficient resources.

> ## Reconsilation Error

Reviewer: ```suggestion
## Reconcilation Error
```

It could happen that the pod appears to be running normally but does not reconcile the resources inside of your Kubernetes cluster.

Check the logs for reconcilation errors:
```
kubectl logs deployment/starboard-operator -n starboard-system
```

> If this is the case, the Starboard Operator likley does not have the right configurations to access your resource. 

Reviewer: ```suggestion
If this is the case, the Starboard Operator likely does not have the right configurations to access your resource. 
```

> ## Operator does not create VulnerabilityReports

Reviewer: ```suggestion
## Operator does not Create VulnerabilityReports
```

> VulnerabilityReports are owned and controlled by the immediate Kubernetes workload. Every VulnerabilityReport of a pod is thus, linked to a [replicaset.](./index.md) In case the Starboard Operator does not create a VulnerabilityReport for your workloads, it could be that it is not monitoring the namespace that your workloads are running on.

Reviewer: ```suggestion
VulnerabilityReports are owned and controlled by the immediate Kubernetes workload. Every VulnerabilityReport of a pod is thus, linked to a [ReplicaSet](./index.md). In case the Starboard Operator does not create a VulnerabilityReport for your workloads, it could be that it is not monitoring the namespace that your workloads are running on.
```

# Troubleshooting the Starboard Operator

> The Starboard operator installs several Kubernetes resources into the `starboard-system` namespace.

Reviewer: Based on our installation docs we demonstrate how to install in in `starboard-system` namespaces. But in practice, you can install it in any namespace. Do you think we can clarify that as well?

Here are the common steps to check whether the operator is running correctly and to troubleshoot common issue.

In addition to having a look at this section, you want to check [previous issues](https://github.com/aquasecurity/starboard/issues) to see if someone from the community had similar problems before.

> Feel free to either [open an issue](https://github.com/aquasecurity/starboard/issues), reach out on [Slack](https://slack.aquasec.com), or post your questions in the [discussion forum.]()

Reviewer: Are we missing the link to discussion forum here?

## Installation

Make sure that the latest version of the Starboard Operator is installed inside of your Kubernetes cluster.
For this, have a look at the installation [options.](./installation)

For instance, if your are using the Helm deployment, you need to check the Helm Chart version deployed to your cluster:
```
helm list -n starboard-system
```

> ## Starboard Pod not running

Reviewer: ```suggestion
## Starboard Pod Not Running
```

> The Starboard Operator will run a pod inside the starboard-system namespace. Make sure that the pod is running correctly:

Reviewer: This is how it looks like when everythings if okey:

```
NAMESPACE            NAME                                         READY   STATUS    RESTARTS      AGE
starboard-system     starboard-operator-6c9bd97d58-hsz4g          1/1     Running   5 (19m ago)   30h
```

```suggestion
The Starboard Operator will run a pod inside the `starboard-system` namespace. Make sure that the pod is in the `Running` status:
```

```
kubectl get pods -n starboard-system
```

> If the pod is in `failed`, `pending`, or `unknown` check the events and the logs of the pod.

Reviewer: We can use uppercase to incline with K8s docs on https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/
```suggestion
If the pod is in `Failed`, `Pending`, or `Unknown` check the events and the logs of the pod.
```

```

> kubectl logs <POD-NAME> -n starboard-system

Reviewer: To retrieve logs you use deployment name instead:

```
kubectl logs deployment/starboard-operator -n starboard-system
```

```

```
kubectl describe pod <POD-NAME> -n starboard-system
```

> Note that you have to replace the pod-name with the name of the starboard pod running inside of your cluster.

Reviewer: ```suggestion
Note that you have to replace the `<POD-NAME>` with the name of the starboard pod running inside of your cluster.
```

If there are too many logs messages, try deleting the Starboard pod and observe its behaviour upon restarting. A new pod should spin up automatically after deleting the failed pod.

## ImagePullBackOff or ErrImagePull

Check the status of the Starboard Operator pod running inside of your Kubernetes cluster. If the Status is ImagePullBackOff or ErrImagePull, it means that the Operator either

* tries to access the wrong image
* cannot pull the image from the registry

Make sure that you are providing the right resources upon installing the Starboard Operator.

## CrashLoopBackOff

If your pod is in `CrashLoopBackOff`, it is likely the case that the pod cannot be scheduled on the Kubernetes node that it is trying to schedule on.
In this case, you want to investigate further whether there is an issue with the node. It could for instance be that the node does not have sufficient resources.

## Reconsilation error

It could happen that the pod appears to be running normally but does not reconcile the resources inside of your Kubernetes cluster.

Check the logs for reconcilation errors:
```

> kubectl logs <POD-NAME> -n starboard-system

Reviewer: See my previous comment about `kubectl logs deploy/starboard-operator -n starboard-system`

```

If this is the case, the Starboard Operator likley does not have the right configurations to access your resource. 

## Operator does not create Vulnerability Reports

> Every VulnerabilityReport is linked to a [deployment.](./index.md) In case the Starboard Operator does not create a VulnerabilityReport for your workloads, it could be that it is not monitoring the namespace that your workloads are running on.

Reviewer: To be precise VulnerabilityReports are owned by (or controlled by) immediate Kubernetes workload. So for deployments this would be active ReplicaSet aka current revision.

> Every VulnerabilityReport is linked to a [deployment.](./index.md) In case the Starboard Operator does not create a VulnerabilityReport for your workloads, it could be that it is not monitoring the namespace that your workloads are running on.

Reviewer 2: `deployment` link is broken ?

Author: it works for me

An easy way to check this is by looking for the ClusterRoleBinding for the Starboard Operator

```

> kubectl get ClusterRoleBinding | grep "starboard-operator"

Reviewer: Or using our great who-can plugin 😄 :

```console
$ kubectl who-can list vulnerabilityreports
No subjects found with permissions to list vulnerabilityreports assigned through RoleBindings

CLUSTERROLEBINDING                           SUBJECT                         TYPE            SA-NAMESPACE
cluster-admin                                system:masters                  Group
starboard-operator                           starboard-operator              ServiceAccount  starboard-system
system:controller:generic-garbage-collector  generic-garbage-collector       ServiceAccount  kube-system
system:controller:namespace-controller       namespace-controller            ServiceAccount  kube-system
system:controller:resourcequota-controller   resourcequota-controller        ServiceAccount  kube-system
system:kube-controller-manager               system:kube-controller-manager  User
```

```

If it does not exist, Starboard currently cannot monitor any namespace outside of the `starboard-system` namespace. 

> For instance, if you are using the [Helm Chart](./installation/helm.md), you want to make sure to set the `targetnamespace` to the namespace that you want to operator to monitor.

Reviewer: IIRC the Helm value is `targetNamespace`

> 
> 
> 
> 
> 
> 

Reviewer: ```suggestion
```
